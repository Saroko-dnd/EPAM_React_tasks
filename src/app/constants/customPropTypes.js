import { createPropType } from 'react-custom-proptypes';

const articlePropType = createPropType(prop => (
    typeof prop === 'object' &&
        typeof prop.author === 'string' &&
        typeof prop.title === 'string' &&
        typeof prop.description === 'string' &&
        typeof prop.url === 'string' &&
        typeof prop.urlToImage === 'string' &&
        typeof prop.publishedAt === 'string' &&
        typeof prop.source === 'object' &&
        typeof prop.source.name === 'string'
), 'Must be news article');

export default { article: articlePropType };
