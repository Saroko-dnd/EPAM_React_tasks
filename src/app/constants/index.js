const newsAppConfig = {
    apiToken: 'e99a027c252b468ab4c0d3d8872f4cfc',
    apiBaseLink: 'https://newsapi.org/v2/',

    navigation: {
        homePage: { destination: '/', title: 'Top news', id: 'nav-link-1' },
        firstPage: {
            destination: '/first',
            title: 'Ultra news',
            id: 'nav-link-2',
        },
        secondPage: {
            destination: '/second',
            title: 'Super news',
            id: 'nav-link-3',
        },
        thirdPage: {
            destination: '/third',
            title: 'Mega news',
            id: 'nav-link-4',
        },
    },

    colors: {
        green: '#009900',
    },
};

export default newsAppConfig;
