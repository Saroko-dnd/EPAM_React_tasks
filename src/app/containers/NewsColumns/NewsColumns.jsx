import React from 'react';
import PropTypes from 'prop-types';

import LoadingIndicator from '../../components/LoadingIndicator';

class NewsColumns extends React.Component {
    componentDidMount() {
        const { newsApiToken, newsApiLink } = this.props;
        const newsApiRequestLink = `${newsApiLink}top-headlines?country=us&apiKey=${newsApiToken}`;
        this.props.fetchNewsData(newsApiRequestLink);
    }

    render() {
        if (this.props.dataUploaded) {
            return (
                <section className="news-columns row">
                    {this.props.columns}
                </section>
            );
        }

        return (
            <section className="news-columns">
                <h2>{}</h2>
                <LoadingIndicator message="Loading news..." />
            </section>
        );
    }
}

NewsColumns.propTypes = {
    newsApiLink: PropTypes.string.isRequired,
    newsApiToken: PropTypes.string.isRequired,
    fetchNewsData: PropTypes.func.isRequired,
    columns: PropTypes.arrayOf(PropTypes.element),
    dataUploaded: PropTypes.bool,
};

NewsColumns.defaultProps = {
    dataUploaded: false,
    columns: [],
};

export default NewsColumns;
