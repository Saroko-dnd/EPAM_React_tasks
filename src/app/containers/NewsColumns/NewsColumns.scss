$color-green: rgb(241, 15, 15);
$loading-font-color: $color-green;

.news-columns{
    .loading-header{
        font-size: 2rem;
        font-weight: bold;
        color: $loading-font-color;
    }
}
