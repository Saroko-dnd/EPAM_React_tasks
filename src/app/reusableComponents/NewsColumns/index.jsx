import React from 'react';

import ArticlesList from '../../reusableComponents/ArticlesList';
import newsAppConfig from '../../configFiles/';
import createUuidv4 from '../../helpers/createUuidv4';

const NewsColumns = () => {
    const columns = [];

    for (let column = 1; column < 5; column += 1) {
        columns.push(<ArticlesList
            className="col-xl-3 col-lg-4 col-md-6"
            key={createUuidv4()}
            newsApiLink={`${
                newsAppConfig.apiBaseLink
            }top-headlines?country=us&pageSize=5&page=${column}&apiKey=${
                newsAppConfig.apiToken
            }`}
        />);
    }

    return <section className="row">{columns}</section>;
};

export default NewsColumns;
