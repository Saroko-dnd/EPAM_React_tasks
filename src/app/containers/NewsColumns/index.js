import _ from 'lodash';
import { connect } from 'react-redux';

import actions from '../../actions';
import NewsColumns from './NewsColumns';
import { actionTypes } from '../../constants';

const mapStateToProps = state => ({
    columns: _.chunk(state.articles, 5),
    dataDownloaded: !state.newsIsLoading,
});

const mapDispatchToProps = dispatch => ({
    loadNews: apiUrl =>
        dispatch(actions.loadNews(apiUrl, actionTypes.TOP_NEWS_DOWNLOADED)),
});

export default connect(mapStateToProps, mapDispatchToProps)(NewsColumns);
