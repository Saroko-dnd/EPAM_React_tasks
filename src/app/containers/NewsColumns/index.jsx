import React from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';

import NewsColumns from './NewsColumns';
import actions from '../../actions';
import ArticlesList from '../../components/ArticlesList';
import constants from '../../constants';

function createColumnsFromNews(articles) {
    const groupedArticles = _.chunk(articles, 5);

    return groupedArticles.map(foundArticles => (
        <ArticlesList
            className="col-xl-3 col-lg-4 col-md-6"
            key={foundArticles[0].url}
            articles={foundArticles}
        />
    ));
}

const mapStateToProps = state => ({
    columns: createColumnsFromNews(state.articles),
    dataUploaded: !state.newsIsLoading,
});

const mapDispatchToProps = dispatch => ({
    fetchNewsData: url =>
        dispatch(actions.loadNews(url, constants.actions.TOP_NEWS_UPLOADED)),
});

export default connect(mapStateToProps, mapDispatchToProps)(NewsColumns);
