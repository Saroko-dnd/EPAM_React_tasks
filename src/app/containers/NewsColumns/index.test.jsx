import NewsColumns from './NewsColumns';
import { actionTypes } from '../../constants';
import { __GetDependency__ } from './index';

describe('NewsColumns container component', () => {
    describe('rendering and behavior', () => {
        const fakeNewsApiToken = 'fake token';
        const fakeNewsApiLink = 'fake link';
        const fakeLoadNews = jest.fn();
        const newsColumnsWrapper = shallow(<NewsColumns
            newsApiToken={fakeNewsApiToken}
            newsApiLink={fakeNewsApiLink}
            loadNews={fakeLoadNews}
        />);

        it('should render loading indicator if property dataDownloaded set to false', () => {
            expect(newsColumnsWrapper).toMatchSnapshot();
        });

        it('should render news columns if property dataDownloaded set to true ', () => {
            newsColumnsWrapper.setProps({
                dataDownloaded: true,
                columns: [
                    <span id="column_1" key="1" />,
                    <span id="column_2" key="2" />,
                    <span id="column_3" key="3" />,
                ],
            });
            expect(newsColumnsWrapper).toMatchSnapshot();
        });
    });

    describe('mapStateToProps', () => {
        const fakeState = {
            articles: [
                { title: 'article_1' },
                { title: 'article_2' },
                { title: 'article_3' },
                { title: 'article_4' },
                { title: 'article_5' },
                { title: 'article_6' },
                { title: 'article_7' },
                { title: 'article_8' },
                { title: 'article_9' },
                { title: 'article_10' },
                { title: 'article_11' },
                { title: 'article_12' },
                { title: 'article_13' },
                { title: 'article_14' },
                { title: 'article_15' },
            ],
            newsIsLoading: false,
        };
        const mapStateToProps = __GetDependency__('mapStateToProps');
        const mappingResult = mapStateToProps(fakeState);

        it('should return object with array of ArticlesList components for news columns', () => {
            expect(mappingResult.columns).toMatchSnapshot();
        });

        it('should return dataDownloaded property with a value opposite to the value of property newsIsLoading in application state', () => {
            expect(mappingResult.dataDownloaded).toEqual(!fakeState.newsIsLoading);
        });
    });

    describe('mapDispatchToProps', () => {
        const fakeDispatch = jest.fn();
        const fakeApiUrl = '/api/top/news';
        const mapDispatchToProps = __GetDependency__('mapDispatchToProps');

        it(`should return object with function loadNews that should dispatch ${
            actionTypes.LOAD_NEWS
        } action with passed api url and${
            actionTypes.TOP_NEWS_DOWNLOADED
        } action type as payload`, () => {
            mapDispatchToProps(fakeDispatch).loadNews(fakeApiUrl);
            expect(fakeDispatch).toBeCalledWith({
                type: actionTypes.LOAD_NEWS,
                payload: {
                    apiLink: fakeApiUrl,
                    downloadedActionType: actionTypes.TOP_NEWS_DOWNLOADED,
                },
            });
        });
    });
});
