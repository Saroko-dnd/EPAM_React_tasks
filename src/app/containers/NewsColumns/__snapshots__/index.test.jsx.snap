// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsColumns container component mapStateToProps should return object with array of ArticlesList components for news columns 1`] = `
Array [
  <ArticlesList
    articles={
      Array [
        Object {
          "title": "article_1",
        },
        Object {
          "title": "article_2",
        },
        Object {
          "title": "article_3",
        },
        Object {
          "title": "article_4",
        },
        Object {
          "title": "article_5",
        },
      ]
    }
    className="col-xl-3 col-lg-4 col-md-6 d-flex flex-column align-items-center justify-content-around"
    linksToDetails={true}
    title=""
  />,
  <ArticlesList
    articles={
      Array [
        Object {
          "title": "article_6",
        },
        Object {
          "title": "article_7",
        },
        Object {
          "title": "article_8",
        },
        Object {
          "title": "article_9",
        },
        Object {
          "title": "article_10",
        },
      ]
    }
    className="col-xl-3 col-lg-4 col-md-6 d-flex flex-column align-items-center justify-content-around"
    linksToDetails={true}
    title=""
  />,
  <ArticlesList
    articles={
      Array [
        Object {
          "title": "article_11",
        },
        Object {
          "title": "article_12",
        },
        Object {
          "title": "article_13",
        },
        Object {
          "title": "article_14",
        },
        Object {
          "title": "article_15",
        },
      ]
    }
    className="col-xl-3 col-lg-4 col-md-6 d-flex flex-column align-items-center justify-content-around"
    linksToDetails={true}
    title=""
  />,
]
`;

exports[`NewsColumns container component rendering and behavior should render loading indicator if property dataDownloaded set to false 1`] = `
<section
  className="news-columns"
>
  <LoadingIndicator
    color="#009900"
    message="Loading news..."
    size={150}
  />
</section>
`;

exports[`NewsColumns container component rendering and behavior should render news columns if property dataDownloaded set to true  1`] = `
<section
  className="news-columns row justify-content-center align-items-stretch"
>
  <span
    id="column_1"
    key="1"
  />
  <span
    id="column_2"
    key="2"
  />
  <span
    id="column_3"
    key="3"
  />
</section>
`;
